add SolrJsonWriter2 using httpx client instead of httpclient

jrochkind opened the pull request:
The default SolrJsonWriter uses [httpclient](https://github.com/nahi/httpclient) for http communications, which has been under-maintained, causing various problems. 

We [have](https://github.com/traject/traject/issues/285#issuecomment-1330950321) discussed using httpx instead. 

This PR introduces an additional writer with a new name, `SolrJsonWriter2`, that uses httpx.  (Suggestions for alternate name welcome).  There is also an ENV variable that can be set, `SOLR_JSON_WRITER=2` to opt-in to the new class being the _default_, for environments where people don't want to make code/settings changes but still opt-in. 

Because some people may be configuring or interacting with the underlying HTTP client directly, I think this is a backwards incompat change, so wanted to introduce it as an additonal alternate opt-in writer, with both writers being available for a given amount of time. 

I liked httpx working with it here. It’s tested under jruby, and has concurrency commitments, including managing an [internal pool of persistent http connections (if so desired) that is thread-safe](https://honeyryderchuck.gitlab.io/httpx/wiki/Connection-Pools). http-rb has also been proposed, see  #304, but does not have this, you have to build it yourself on top.  httpclient does have this internal thread-safe persistent pool, and at least in the past we believed it was important for performance in some configurations.  Of course, there can be different writers available as a choices, but I think httpx is good for future default. 

Hopefully a few people would try this new writer. After it’s been verified a bit in the field, we would change it to be the new default — i would lean towards this being a major version release, but would be open to it being a minor version release if people prefer — with the old one initially still being available, in case anyone runs into a problem and wants to switch back to it.  At that point `SolrJsonWriter` might point to the new one, I dunno the old one might be `SolrJsonWriterOld`.

I am very happy to discuss this plan and make alterations based on any feedback.  Including names of classes to try to make this less confusing. 

@billdueber ?

jrochkind committed:
add SolrJsonWriter2 using httpx client instead of httpclient

jrochkind committed:
ENV var can opt into SolrJsonWriter2 default

jrochkind committed:
doc default writer changes coming, and opt-in to SolrJsonWriter2

jrochkind commented:
Aha, the latest httpx requires ruby 2.7.0+

Jruby 9.3.0 is considered ruby 2.6. compatible, but not considered ruby 2.7 compat. So bundler refuses to install with httpx 1.4 dependency and jruby 9.3.

And jruby-9.4 remains failing for unrelated namespace reasons?

If this general approach is approved, we can consider what to do, with options including 
* dropping jruby-9.3 support
* at least for now, making httpx a "soft" opt-in dependency not in the gemspec, so it doesn't prevent the gem being installed with jruby-9.3. 
* Relaxing httpx dependency to allow older versions (but the latest has changed semantics for connection pooling, I thought it better to stick with only those new ones)

```
Bundler found conflicting requirements for the Ruby version:
  In Gemfile:
    Ruby

    bundler (>= 1.[7](https://github.com/traject/traject/actions/runs/13290792500/job/37110843236?pr=307#step:5:8), < 3) was resolved to 2.2.33, which depends on
      Ruby (>= 2.3.0)

    traject was resolved to 3.[8](https://github.com/traject/traject/actions/runs/13290792500/job/37110843236?pr=307#step:5:9).3, which depends on
      httpx (~> 1.4) was resolved to 1.4.0, which depends on
        Ruby (>= 2.7.0)

    minitest was resolved to 5.25.4, which depends on
      Ruby (< 4.0, >= 2.6)

    traject was resolved to 3.8.3, which depends on
      nokogiri (~> 1.[9](https://github.com/traject/traject/actions/runs/13290792500/job/37110843236?pr=307#step:5:10)) was resolved to 1.13.[10](https://github.com/traject/traject/actions/runs/13290792500/job/37110843236?pr=307#step:5:11), which depends on
        Ruby (>= 2.6.0)

    rspec-mocks (~> 3.4) was resolved to 3.[13](https://github.com/traject/traject/actions/runs/13290792500/job/37110843236?pr=307#step:5:14).2, which depends on
      Ruby (>= 1.8.7)

    webmock (~> 3.4) was resolved to 3.25.0, which depends on
      Ruby (>= 2.5)
```

billdueber commented:
Are you sure we want the `ENV[whatever_it_was] == 2`? It feels like adding another layer of specificity when it's no harder to just set the class name.

jrochkind commented:
@billdueber I know there were some people who really wanted to make this the new default. 

I don't think it should be the new default until after a period of some overlap. 

So i tried to think of anything I could do to get as close to what those folks wanted as possible. My thinking was there might be some environments where people don't want to change their configuration files to point to a new writer (that they may have to/want to change back later when the default changes again).  Perhaps they don't want to make commits to repo that holds config, or have something in a container they don't want to change (although I guess both of those cases you have to chang *something* to get the EN V anyway), or the config files are maintained by metadata technicians that they want to shield from this implementation detail. 

But that's not a use case I have/need, it was just me trying to be hospitable. So if we can't find anyone who actually says they want this, and you aren't fond of it, we can take it out. 

Curious if @jcoyne has any input.

jrochkind commented:
Hi, I may merge this myself with no review, if there are no other maintainers with capacity to review! 

@billdueber if you would approve if I remove the ENV thing, i will do so!  Or any other feedback you have.

jrochkind committed:
fix tests mocking too much giving false positive, and fix func with hidden breakage

Errors weren't actually being tested, along with some other func. We mocked too much when we switched from HttpClient. Switch to WebMock, better level of abstraction, feel more confident in tests.

billdueber commented:
The Nokogiri namespace stuff is kinda killing me. I'll take a look at it unless someone knows for sure what's going on. I remember dealing with it before, but don't remember what the deal was, other than recalling that it had to do (I think) with how the tests are set up. Ringing bells for anyone?

billdueber commented:
Oh, here it is. My "solution" at the time was, apparently, to just not run the namespace tests under jruby. https://github.com/traject/traject/compare/main...ignore_jruby_namespace_errors

jrochkind commented:
I am fully prepared to merge anyway with failing jruby. 🤷  (not caused by this PR, pre-existing). 

Maybe make another issue to track jruby failures if someone is actually interested in dealing with them. Yeah, it's annoying. I haven't used jruby for years myself.

jrochkind committed:
change name of new SolrJsonWriter2 to SolrJsonHttpxWriter

less confusing in the long run.

jrochkind committed:
test basic auth with actual mock http instead of private api, so works on all httpx versions
